tfe/provider.go

```diff
@@ -0,0 +1,194 @@
+	for userHost, hostConfig := range config.Hosts {
+		host, err := svchost.ForComparison(userHost)
+		if err != nil {
+			// ignore invalid hostnames.
+			continue
+		}
+		services.ForceHostServices(host, hostConfig.Services)
+	}
+
+	// Discover the full Terraform Enterprise service address.
+	address := services.DiscoverServiceURL(host, serviceID)
+	if address == nil {
+		return nil, fmt.Errorf("host %s does not provide a Terraform Enterprise API", host)
+	}
+
+	// Get any configured credentials from the CLI Config File.
+	creds, err := services.CredentialsForHost(host)
+	if err != nil {
+		log.Printf("[DEBUG] Failed to get credentials for %s: %s (ignoring)", host, err)
+	}
+	if creds != nil {
+		token = creds.Token()
+	}
+
+	// Create a new TFE client config..
+	cfg := &tfe.Config{
+		Address: address.String(),
+		Token:   token,
+	}
+
+	// Create s new TFE client.
+	return tfe.NewClient(cfg)
+}
+
+// cliConfig tries to find and parse the configuration of the Terraform CLI.
+// This is an optional step, so any errors are ignored.
+func cliConfig() *Config {
+	config := &Config{}
+
+	// Detect the CLI config file path.
+	configFilePath := os.Getenv("TERRAFORM_CONFIG")
+	if configFilePath == "" {
+		filePath, err := configFile()
+		if err != nil {
+			log.Printf("[ERROR] Error detecting default CLI config file path: %s", err)
+			return config
+		}
+		configFilePath = filePath
+	}
+
+	// Read the CLI config file content.
+	content, err := ioutil.ReadFile(configFilePath)
+	if err != nil {
+		log.Printf("[ERROR] Error reading the CLI config file: %v", err)
+		return config
+	}
+
+	// Parse the CLI config file content.
+	obj, err := hcl.Parse(string(content))
+	if err != nil {
+		log.Printf("[ERROR] Error parsing the CLI config file: %v", err)
+		return config
+	}
+
+	// Decode the CLI config file content.
+	if err := hcl.DecodeObject(&config, obj); err != nil {
+		log.Printf("[ERROR] Error decoding the CLI config file: %v", err)
+	}
+
+	return config
+}
+
+func credentialsSource(config *Config) auth.CredentialsSource {
+	creds := auth.NoCredentials
+
+	// Add all configured credentials to the credentials source.
+	if len(config.Credentials) > 0 {
+		staticTable := map[svchost.Hostname]map[string]interface{}{}
+		for userHost, creds := range config.Credentials {
+			host, err := svchost.ForComparison(userHost)
+			if err != nil {
+				// We expect the config was already validated by the time we get
+				// here, so we'll just ignore invalid hostnames.
+				continue
+			}
+			staticTable[host] = creds
+		}
+		creds = auth.StaticCredentialsSource(staticTable)
+	}
+
+	return creds
+}
+
+var descriptions = map[string]string{
+	"hostname": "The remote backend hostname to connect to. Default to app.terraform.io.",
+	"token": "The token used to authenticate with the remote backend. If `TFE_TOKEN` is set\n" +
+		"or credentials for the host are configured in the CLI Config File, then those\n" +
```

Comment:
afaik in-line specified values override environment variables

Reply:
Yeah, I will indeed make sure any value declared in the provider configuration takes precedence over the env variable and the CLI config file. So it will be `inline` > `env var` > `CLI config file`.

Fixed in the next commit.